### DISCUSSION TITLE
不Delete Channel是否会造成内存泄漏的问题

### DISCUSSION THREAD

**Discussion by guang19** [chenshuo/muduo#616] [General]
大神，我发现TcpConnection的CloseCallback最终执行的是connectDestroyed：
```
void TcpConnection::connectDestroyed()
{
  loop_->assertInLoopThread();
  if (state_ == kConnected)
  {
    setState(kDisconnected);
    channel_->disableAll();

    connectionCallback_(shared_from_this());
  }
  channel_->remove();
}
```
但这个函数仅仅是把channel从Poller和ChannelMap中移除，并不会delete Channel，我也没在其它任何地方找到delete  Channel相关的代码，不知道这是否是一个内存泄漏问题？

**Comment by chenshuo** [comment]
1. 你可以在 `Channel::~Channel()` 打一下 log，然后创建并断开几个连接，看看它有没有析构。
2. 其实整个 `muduo_net` 库只有一处不得已用了 `delete`，是在 `class TimerQueue` 里。在源码里你也找不的 `TcpConnection`/`Socket` 的 `delete` 语句。

**Reply by guang19** [reply]
好的，谢谢前辈指教。

**Reply by guang19** [reply]
啊，我傻了，居然没看到是使用的智能指针管理Channel了
